File: android/app/src/main/java/org/castaliainstitute/aethertak/field/media/MediaIntegrity.kt
package org.castaliainstitute.aethertak.field.media

import java.io.InputStream
import java.security.MessageDigest

data class MediaFileIntegrity(
    val sha256: String,
    val sizeBytes: Long,
)

fun inspectMedia(input: InputStream): MediaFileIntegrity {
    val digest = MessageDigest.getInstance("SHA-256")
    val buffer = ByteArray(1024 * 1024)
    var sizeBytes = 0L
    while (true) {
        val count = input.read(buffer)
        if (count < 0) break
        if (count == 0) continue
        require(sizeBytes <= Long.MAX_VALUE - count) {
            "The captured media file is too large to inspect safely."
        }
        digest.update(buffer, 0, count)
        sizeBytes += count
    }
    require(sizeBytes > 0) { "The captured media file is empty." }
    return MediaFileIntegrity(
        sha256 = digest.digest().joinToString("") { "%02x".format(it) },
        sizeBytes = sizeBytes,
    )
}

File: ios/App/App/MediaIntegrity.swift
import CryptoKit
import Foundation

struct MediaFileIntegrity: Equatable {
    let sha256: String
    let sizeBytes: Int64
}

enum MediaIntegrityError: LocalizedError {
    case emptyFile
    case invalidSize

    var errorDescription: String? {
        switch self {
        case .emptyFile:
            return "The captured media file is empty."
        case .invalidSize:
            return "The captured media file is too large to inspect safely."
        }
    }
}

func inspectMediaFile(at url: URL) throws -> MediaFileIntegrity {
    let handle = try FileHandle(forReadingFrom: url)
    defer { try? handle.close() }

    var digest = SHA256()
    var sizeBytes: Int64 = 0
    while true {
        let data = try handle.read(upToCount: 1024 * 1024) ?? Data()
        if data.isEmpty { break }
        guard sizeBytes <= Int64.max - Int64(data.count) else {
            throw MediaIntegrityError.invalidSize
        }
        sizeBytes += Int64(data.count)
        digest.update(data: data)
    }
    guard sizeBytes > 0 else {
        throw MediaIntegrityError.emptyFile
    }

    return MediaFileIntegrity(
        sha256: digest.finalize()
            .map { String(format: "%02x", $0) }
            .joined(),
        sizeBytes: sizeBytes
    )
}
